File: src/facond/actions.js
import debug from 'debug'

var log = debug('facond.action')

/**
 * Apply a reversible action on the DOM if conditions apply, unapply otherwise.
 */
class Action {
  /**
  * @param conditions List of :js:cls:`Condition` objects
  */
  constructor(conditions) {
    this.conditions = conditions || []
  }

  /**
   * Call :js:meth:`apply()` if conditions validate, :js:meth:`unapply` otherwise.
   *
   * @param form :js:class:`Form` object to execute on.
   */
  execute(form) {
    let method = 'apply'

    for (var i=0; i<this.conditions.length; i++) {
      if (!this.conditions[i].validate(form)) {
        method = 'unapply'
        break
      }
    }

    log(this, '.', method, '(', this.field, ')')
    this[method](form)
  }

  /**
   * Modify the DOM, should save the state for :js:meth:`unapply()`.
   *
   * @param form :js:class:`Form` object to apply on.
   */
  apply(form) { // eslint-disable-line no-unused-vars
    console.log('not implemented')
  }

  /**
   * Restore the DOM prior to :js:meth:`apply()` call.
   *
   * @param form :js:class:`Form` object to unapply on.
   */
  unapply(form) { // eslint-disable-line no-unused-vars
    console.log('not implemented')
  }
}

/**
 * Remove a field from a form.
 */
class RemoveField extends Action {
  /**
  * @param conditions List of :js:class:`Condition` objects
  * @param :js:class:`Field` Subject Field instance.
  */
  constructor(conditions, field) {
    super(conditions)
    this.field = field
  }

  // Hide the field.
  apply(form) {
    form.field(this.field).hide()
  }

  // Show the field.
  unapply(form) {
    form.field(this.field).show()
  }
}


class ChoicesAction extends Action {
  /**
  * @param conditions List of :js:class:`Condition` objects
  */
  constructor(conditions, field, choices) {
    super(conditions)
    this.field = field
    this.choices = choices
  }

  // Hide options which are not in this.choices from a field.
  apply(form) {
    let field = form.field(this.field)

    if (!this.validate(field.value)) {
      field.valueReset()
    }

    for (let i=0; i < field.element.options.length; i++) {
      if (this.validate(field.element.options[i].value)) {
        field.element.options[i].classList.remove('facond-hide')
      } else {
        let option = field.element.options[i]
        option.classList.add('facond-hide')
        option.selected = false
      }
    }

    if (!field.multiple) {
      // If selected value was removed, empty the field
      if (!this.validate(field.value)) {
        let empty = field.element.querySelector('option[value=""]')
        if (empty === undefined) {
          field.element.prepend('<option value=""></option>')
        }
        field.value = ''
      }
    }
  }
}


/**
 * Remove given choices from a field.
 */
class RemoveChoices extends ChoicesAction {
  validate(value) {
    return this.choices.indexOf(value) < 0
  }

  // Show options which are not in this.choices from a field.
  unapply(form) {
    let  field = form.field(this.field)

    for (let i=0; i < field.element.options.length; i++) {
      if (!this.choices.indexOf(field.element.options[i].value)) {
        field.element.options[i].classList.remove('facond-hide')
      }
    }
  }
}

class SetChoices extends ChoicesAction {
  validate(value) {
    return this.choices.indexOf(value) >= 0
  }
}

export {
  Action,
  ChoicesAction,
  RemoveChoices,
  SetChoices,
  RemoveField,
}

File: src/facond/actions.py
"""
An Action may apply reversible mutations on :doc:`Form<forms>`.

It should always take a list of :doc:`Condition<conditions>` as first
argument, and then whatever it's going to need to work.

On the server side,
"""
from collections import OrderedDict

from django import forms

from .js import JsDictMixin


class Action(JsDictMixin):
    """An action with an optional list of :py:cls:`~facond.forms.Form`.

    The :py:meth:`~execute()` will first execute
    :py:meth:`~facond.conditions.Condition.validate()` for every of its
    conditions against the :py:cls:`~facond.forms.Form` instance it is passed.
    """

    js_attrs = ['field', 'conditions']

    def __init__(self, conditions):
        """Instanciate with a list of condition to require."""
        self.conditions = conditions

    def execute(self, form):
        """Prevent the action from being applied if conditions don't pass."""
        for condition in self.conditions:
            if not condition.validate(form):
                return

        self.apply(form)
        return self

    def apply(self, form):
        """Actual application of the action on the form field."""
        raise NotImplementedError()

    def unapply(self, form):
        """Revert apply() action."""


class RemoveField(Action):
    """Remove fields from a form."""

    def __init__(self, conditions, field):
        """Remove a field if conditions validate."""
        super(RemoveField, self).__init__(conditions)
        self.field = field

    def apply(self, form):
        """Remove the field and data from field.form."""
        form[self.field].extra_css_classes = 'facond-hide'
        self.original_keys = list(form.fields.keys())
        self.original_field = form.fields.pop(self.field)

    def unapply(self, form):
        """Restore the field and data in field.form."""
        original_field = getattr(self, 'original_field', None)
        if not original_field:
            return
        fields = OrderedDict()
        for key in self.original_keys:
            if key == self.field:
                fields[key] = original_field
            elif key in form.fields:
                fields[key] = form.fields[key]
        form.fields = fields


class ChoicesAction(Action):
    js_attrs = ['field', 'choices', 'conditions']

    def __init__(self, conditions, field, choices):
        """List of choice values to remove if conditions pass."""
        super().__init__(conditions)
        self.field = field
        self.choices = choices

    def apply(self, form):
        """Raise a ValidationError if the field value is in self.values."""
        values = form.data.get(self.field, None)
        if not isinstance(values, list):
            values = [values]

        for value in values:
            if not self.validate(value):
                raise forms.ValidationError(
                    form.fields[self.field].error_messages['invalid_choice'],
                    code='invalid_choice',
                    params={'value': value},
                )


class RemoveChoices(ChoicesAction):
    """Remove choices from a choice field."""

    def validate(self, value):
        return value not in self.choices


class SetChoices(ChoicesAction):
    """Remove choices from a choice field."""

    def validate(self, value):
        return value in self.choices
